lib/chef-dk/policyfile/undo_stack.rb

```diff
@@ -0,0 +1,98 @@
+#
+# Copyright:: Copyright (c) 2015 Chef Software Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'fileutils'
+
+require 'ffi_yajl'
+
+require 'chef-dk/helpers'
+require 'chef-dk/policyfile/undo_record'
+
+module ChefDK
+  module Policyfile
+
+    class UndoStack
+
+      MAX_SIZE = 10
```

Comment:
Yeah, here's the design work I did before writing this: https://gist.github.com/danielsdeleo/1723952c9175a3e80c10

Basically, I see 2 kinds of interaction models. For cookbook artifacts and policy revisions, it's totally possible to determine which ones are not in use and clean them up via a garbage collection strategy. In those cases you don't really need a backup or undo (I think) because you know they're unused. For policy groups, I don't think we can come up with a solid and universal way to distinguish "garbage" from "non-garbage," so you need a human to say "delete this thing." But humans make mistakes so we want undo functionality to help humans recover from mistakes.